client/storage.c

```diff
@@ -193,6 +193,304 @@ static int sort_callback(const void *l, const void *r)
+	
+	while (start_time < end_time) {
+		sprintf(path, STORAGE_FILE, STORAGE_FILE_ARGS(start_time));
+				
+		pthread_mutex_lock(&storage_mutex);
+		FILE *f = xdag_open_file(path, "rb");
+		if (f) {
+			if (fseek(f, pos, SEEK_SET) < 0) todo = 0;
+			else todo = fread(readbuf, sizeof(struct xdag_block), bufsize, f);
+			xdag_close_file(f);
+		} else {
+			todo = 0;
+		}
+		pthread_mutex_unlock(&storage_mutex);
+		
+		if(todo > 0) {
+			for (int i = 0; i < todo; ++i) {
+				s.size += sizeof(struct xdag_block);
+				for (int j = 0; j < sizeof(struct xdag_block) / sizeof(uint64_t); ++j) {
+					s.sum += ((uint64_t*)(readbuf + i))[j];
+				}
+			}
+			
+			struct xdag_block *bs = (struct xdag_block*)malloc(sizeof(struct xdag_block)*todo);
+			memcpy(bs, readbuf, sizeof(struct xdag_block)*todo);
+			
+			while (simple_queue_push(bs, sizeof(struct xdag_block)*todo/*, path*/) == 0) {
+#if defined (_WIN32) || defined (_WIN64)
+				sleep(1);
+#else
+				sleep(.001);
+#endif
+			}
+		}
+		
+		if (todo != bufsize) {
+			if (f) {
+				pthread_mutex_lock(&storage_mutex);
+				
+				int res = correct_storage_sums(start_time, &s, 0);
+				
+				pthread_mutex_unlock(&storage_mutex);
+				
+				if (res) {
+					break;
+				}
+				
+				s.size = s.sum = 0;
+				mask = (1l << 16) - 1;
+			} else if (sprintf(path, STORAGE_DIR3, STORAGE_DIR3_ARGS(start_time)), xdag_file_exists(path)) {
+				mask = (1l << 16) - 1;
+			} else if (sprintf(path, STORAGE_DIR2, STORAGE_DIR2_ARGS(start_time)), xdag_file_exists(path)) {
+				mask = (1l << 24) - 1;
+			} else if (sprintf(path, STORAGE_DIR1, STORAGE_DIR1_ARGS(start_time)), xdag_file_exists(path)) {
+				mask = (1ll << 32) - 1;
+			} else {
+				mask = (1ll << 40) - 1;
+			}
+			
+			start_time |= mask;
+			start_time++;
+			
+			pos = 0;
+		} else {
+			pos += todo;
+		}
+	}
+	
+	for (int i = 0; i < QUEUE_SIZE; i++) {
+		g_msg_queue[i].status = QUEUE_DONE;
+	}
+	
+	return 0;
+}
+
+static void *init_storage_add_block_thread(void *data)
+{
+	struct init_storage_param *param = (struct init_storage_param *)data;
+	struct queue_item *item = NULL;
+	struct xdag_block *pbuffer[bufsize];
+	struct xdag_storage_sum s;
+	
+	uint64_t sum = 0, pos = 0;
+	int64_t i, j, k;
+	
+	while (1) {
+		int finish = 1;
+		for (int ii = 0; ii < QUEUE_SIZE; ii++) {
+			if (!(g_msg_queue[ii].status & QUEUE_DONE && g_msg_queue[ii].status & QUEUE_WRITE)) {
+				finish = 0;
+			}
+		}
+		if(finish) {
+			break;
+		}
+		
+		item = simple_queue_splice();
+		if(!item) {
+#if defined (_WIN32) || defined (_WIN64)
+			sleep(1);
+#else
+			sleep(.001);
+#endif
+			continue;
+		}
+				
+		uint64_t todo = item->size / sizeof(struct xdag_block);
+		uint64_t pos0 = pos;
+		
+//		memcpy(buffer, item->data, item->size);
+		struct xdag_block *buffer = item->data;
+		
+		for (i = k = 0; i < todo; ++i, pos += sizeof(struct xdag_block)) {
+			s.size += sizeof(struct xdag_block);
+//			if (buf[i].field[0].time >= start_time && buf[i].field[0].time < end_time) {
```

Comment:
The same - is this line was temporary commented of should be removed?

Reply:
The same reason, I just remove the time range check when load blocks, suppose the blocks are correct. When finish testing, will remove this line. Thank you.